File: frontend/src/lib/archive.ts
import { api } from "@/lib/api";

export type ArchiveDashboard = {
  course_count: number;
  video_count: number;
  published_video_count: number;
  draft_video_count: number;
};

export type ArchiveCourse = {
  id: number;
  name: string;
  is_published: boolean;
  website_published: boolean;
  video_count: number;
  website_url: string | null;
};

export type ArchiveVideo = {
  id: number;
  name: string;
  url: string | null;
  video_url: string | null;
  youtube_id: string | null;
  description: string | null;
  course: [number, string] | null;
  is_published: boolean;
  website_published: boolean;
  website_url: string | null;
  tags: unknown[];
};

export type YouTubePreview = {
  url: string;
  video_id: string;
  title: string | null;
  author_name: string | null;
  thumbnail_url: string | null;
  provider_name: string | null;
  fetched: boolean;
};

export async function login(password: string) {
  const response = await api.post<{ access_token: string }>("/api/v1/auth/login", { password });
  return response.data;
}

export async function fetchDashboard() {
  const response = await api.get<ArchiveDashboard>("/api/v1/archive/dashboard");
  return response.data;
}

export async function fetchCourses() {
  const response = await api.get<ArchiveCourse[]>("/api/v1/archive/courses");
  return response.data;
}

export async function fetchVideos() {
  const response = await api.get<ArchiveVideo[]>("/api/v1/archive/youtube/videos");
  return response.data;
}

export async function previewYouTube(url: string) {
  const response = await api.post<YouTubePreview>("/api/v1/archive/youtube/preview", { url });
  return response.data;
}

export async function createYouTubeVideo(payload: {
  url: string;
  course_name: string;
  title: string;
  description: string;
  publish: boolean;
  tags: string[];
}) {
  const response = await api.post<ArchiveVideo>("/api/v1/archive/youtube/videos", payload);
  return response.data;
}

File: backend/app/schemas/archive.py
from typing import Any

from pydantic import BaseModel, Field


class ArchiveCourse(BaseModel):
    id: int
    name: str
    is_published: bool = False
    website_published: bool = False
    video_count: int = 0
    website_url: str | None = None


class ArchiveCourseCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    publish: bool = False


class YouTubePreviewRequest(BaseModel):
    url: str = Field(min_length=8)


class YouTubePreview(BaseModel):
    url: str
    video_id: str
    title: str | None = None
    author_name: str | None = None
    thumbnail_url: str | None = None
    provider_name: str | None = None
    fetched: bool = False


class YouTubeVideoCreate(BaseModel):
    url: str = Field(min_length=8)
    course_name: str = Field(min_length=2, max_length=160)
    title: str = Field(min_length=2, max_length=220)
    description: str = ""
    publish: bool = False
    tags: list[str] = Field(default_factory=list)


class YouTubeVideoUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=220)
    description: str | None = None
    publish: bool | None = None
    tags: list[str] | None = None


class ArchiveVideo(BaseModel):
    id: int
    name: str
    url: str | None = None
    video_url: str | None = None
    youtube_id: str | None = None
    description: str | None = None
    course: tuple[int, str] | None = None
    is_published: bool = False
    website_published: bool = False
    website_url: str | None = None
    tags: list[Any] = Field(default_factory=list)


class ArchiveDashboard(BaseModel):
    course_count: int
    video_count: int
    published_video_count: int
    draft_video_count: int
